### PR TITLE
Implementing custom PathClean function to allow masking, closes #1298

### LINKED ISSUE #1298: Multiple slashes become one slash
### 1. What version of Caddy are you running (`caddy -version`)?
Caddy (untracked dev build)
From `go get github.com/mholt/caddy/caddy`.
On 5:30 PM (UTC) 16 December 2016.

### 2. What are you trying to do?
Confirm bug exists even on master branch of Caddy after discovering it on Caddy 0.9.3.

### 3. What is your entire Caddyfile?
```
localhost {
  log
}
```

### 4. How did you run Caddy (give the full command and describe the execution environment)?
On macOS 10.12.2, create the above Caddyfile in `$GOPATH/bin`, run `./caddy` after `cd $GOPATH/bin` in the terminal, and go to http://localhost:2015/path/https://www.google.com in the browser.

### 5. What did you expect to see?
I expect to see `GET /path/https://www.google.com HTTP/1.1` in `$GOPATH/bin/access.log`.

### 6. What did you see instead (give full error messages and/or log)?
I see `GET /path/https:/www.google.com HTTP/1.1` in `$GOPATH/bin/access.log`.

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?
Go through the above.

### PR DESCRIPTION
**Pull request by ibnesayeed**
* A generic function is implemented to allow masking of certain patterns before calling the built-in `path.Clean()` function.
* Added default mask to preserve double slashes of protocol separator (e.g., `http://`, `https://`, or `ftp://`).
* Unit tests for the new functionality.

Note: This PR implements the functionality discussed in #1298.

### CODE REVIEW

File: caddyhttp/httpserver/pathcleaner.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+//   - collapse multiple slashes into one
+//   - eliminate "/." (current directory)
+//   - eliminate "<parent_directory>/.."
+// by masking certain patterns in the path with a temporary random string.
+// This could be helpful when certain patterns in the path are desired
+// that would otherwise be changed in the path clean up process.
+// One such use case is the presence of the double slashes as protocol separator
+// (e.g., /api/endpoint/http://example.com).
+// This is a common pattern in many applications to allow passing URIs as path argument
+func CleanMaskedPath(p string, mask ...string) string {
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: Can you choose more descriptive variable names? Rename `p` to `reqPath`, for example, and `mask` to `masks`...

File: caddyhttp/httpserver/pathcleaner.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	var t string
+	maskMap := make(map[string]string)
+
+	// Iterate over supplied masks and create temporary replacement strings
+	// only for the masks that are present in the path, then replace all occurrences
+	for _, m := range mask {
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: Then rename `m` to `mask`.

File: caddyhttp/httpserver/pathcleaner.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+//   - eliminate "/." (current directory)
+//   - eliminate "<parent_directory>/.."
+// by masking certain patterns in the path with a temporary random string.
+// This could be helpful when certain patterns in the path are desired
+// that would otherwise be changed in the path clean up process.
+// One such use case is the presence of the double slashes as protocol separator
+// (e.g., /api/endpoint/http://example.com).
+// This is a common pattern in many applications to allow passing URIs as path argument
+func CleanMaskedPath(p string, mask ...string) string {
+	var t string
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: Rename `t` to `tmpString` or `replacerVal` or something.

File: caddyhttp/httpserver/pathcleaner.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package httpserver
+
+import (
+	"math/rand"
+	"path"
+	"strings"
+	"time"
+)
+
+// A proxy function to prevent one or more of the path cleanup operations:
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: What is a "proxy" function? Revise comment to conform to godoc conventions: "CleanMaskedPath prevents ..."

File: caddyhttp/httpserver/pathcleaner.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+//   - collapse multiple slashes into one
+//   - eliminate "/." (current directory)
+//   - eliminate "<parent_directory>/.."
+// by masking certain patterns in the path with a temporary random string.
+// This could be helpful when certain patterns in the path are desired
+// that would otherwise be changed in the path clean up process.
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: "changed by path.Clean()"

File: caddyhttp/httpserver/pathcleaner.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+// One such use case is the presence of the double slashes as protocol separator
+// (e.g., /api/endpoint/http://example.com).
+// This is a common pattern in many applications to allow passing URIs as path argument
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: End godoc sentences with a period. ;)

File: caddyhttp/httpserver/pathcleaner.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	p = path.Clean(p)
+
+	// Revert the replaced masks after path cleanup
+	for m, t := range maskMap {
+		p = strings.Replace(p, t, m, -1)
+	}
+	return p
+}
+
+func CleanPath(p string) string {
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: Needs godoc comment. Maybe move the comment below up to godoc here.

File: caddyhttp/httpserver/pathcleaner.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	// Apply the default mask to preserve double slashes of protocols
+	// such as "http://", "https://", and "ftp://" etc.
+	return CleanMaskedPath(p, "://")
+}
+
+// The most efficient method for random string generation.
+// Inspired by http://stackoverflow.com/a/31832326.
+const randomStringLenght = 10
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: "randomStringLength"

Also reduce to 4 or 5, I don't think we need a whole 10.

File: caddyhttp/httpserver/pathcleaner_test.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+func TestPreserveDots(t *testing.T) {
+	maskedTestRunner(t, "preserve_dots", "/..", "/.")
+}
+
+func TestDefaultMask(t *testing.T) {
+	for p, v := range paths {
+		assertEqual(t, v["preserve_protocol"], CleanPath(p))
+	}
+}
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner_test.go]: You can add the benchmark tests if you want -- might as well, you've already written them!

**Comment by ibnesayeed** [comment]
> @mholt
> Thanks for submitting! Mostly details. :)

All taken care I guess! :-)

File: caddyhttp/httpserver/pathcleaner.go (hunk 4 above)
Comment (ibnesayeed) [inline caddyhttp/httpserver/pathcleaner.go]: Now you know how "limited" my knowledge is when it comes to golang. I never read the godoc conventions before receiving this comment. ;-)

File: caddyhttp/httpserver/pathcleaner.go (hunk 1 above)
Comment (ibnesayeed) [inline caddyhttp/httpserver/pathcleaner.go]: As they say, naming things is one of the only two hard things in Computer Science. :-)

File: caddyhttp/httpserver/pathcleaner.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+// CleanPath calls CleanMaskedPath() with the default mask of "://"
+// to preserve double slashes of protocols
+// such as "http://", "https://", and "ftp://" etc.
+func CleanPath(reqPath string) string {
+	return CleanMaskedPath(reqPath, "://")
+}
+
+// An efficient and fast method for random string generation.
+// Inspired by http://stackoverflow.com/a/31832326.
+const randomStringLenght = 4
```
Comment (mholt) [inline caddyhttp/httpserver/pathcleaner.go]: This is still spelled wrong. :smile: `randomStringLength`

Comment (ibnesayeed) [inline caddyhttp/httpserver/pathcleaner.go]: Oops! I didn't realize that you pointed a typo out in the first place. I thought you only talked about reducing the length. ;-)

**Comment by ibnesayeed** [comment]
@mholt is there anything else stopping it from being merged? Please let me know.

**Comment by mholt** [comment]
Thanks @ibnesayeed! Would you like to contribute more to the project via pull requests, issues, and reviews? If so I can make you a collaborator so you can push to a branch instead of to a fork.

**Comment by ibnesayeed** [comment]
Thanks for the offer @mholt. I won't mind being in the collaborators of the project. However, I would note that I will be little busy this upcoming year in my graduation, so my contributions will be limited. It was really fun to work on this issue and you were a great reviewer.

**Comment by mholt** [comment]
Fair enough -- feel free to contribute again any time you can. :+1: Good luck on your graduation!

**Pull request merged**